modules/caddyhttp/server.go

```diff
@@ -703,6 +664,61 @@ func (s *Server) shouldLogRequest(r *http.Request) bool {
 	return !s.Logs.SkipUnmappedHosts
 }
 
+// logRequest logs the request to access logs, unless skipped.
+func (s *Server) logRequest(
+	accLog *zap.Logger, r *http.Request, wrec ResponseRecorder, duration *time.Duration,
+	repl *caddy.Replacer, bodyReader *lengthReader, shouldLogCredentials bool,
+) {
+	// this request may be flagged as omitted from the logs
+	if skipLog, ok := GetVar(r.Context(), SkipLogVar).(bool); ok && skipLog {
+		return
+	}
+
+	repl.Set("http.response.status", wrec.Status()) // will be 0 if no response is written by us (Go will write 200 to client)
+	repl.Set("http.response.size", wrec.Size())
+	repl.Set("http.response.duration", duration)
+	repl.Set("http.response.duration_ms", duration.Seconds()*1e3) // multiply seconds to preserve decimal (see #4666)
+
+	logger := accLog
+	if s.Logs != nil {
+		logger = s.Logs.wrapLogger(logger, r.Host)
+	}
+
+	log := logger.Info
+	if wrec.Status() >= 400 {
+		log = logger.Error
+	}
+
+	userID, _ := repl.GetString("http.auth.user.id")
+
+	reqBodyLength := 0
+	if bodyReader != nil {
+		reqBodyLength = bodyReader.Length
+	}
+
+	extra := r.Context().Value(ExtraLogFieldsCtxKey).(*ExtraLogFields)
+
+	// preallocate with the amount of known fields
+	fieldCount := 6
+	fields := make([]zapcore.Field, fieldCount+len(extra.fields))
+	fields[0] = zap.Int("bytes_read", reqBodyLength)
+	fields[1] = zap.String("user_id", userID)
+	fields[2] = zap.Duration("duration", *duration)
+	fields[3] = zap.Int("size", wrec.Size())
+	fields[4] = zap.Int("status", wrec.Status())
+	fields[5] = zap.Object("resp_headers", LoggableHTTPHeader{
```

Comment:
This could probably become a direct assignment:

```go
fields := []zapcore.Field{field1, field2, ...}
```

I think that has the same result and nicer ergonomics.

Reply:
Oh, I see why it's done this way. How about:

```
fields := make([]zapcore.Field, 0, fieldCount+len(extra.fields))
```

and then use append -- is that equally good? Might be simpler code.

Reply:
Got it :+1: